### PR TITLE
Add Support For Batched Queries

### LINKED ISSUE #185: Support Batched Queries
I'd like to submit a PR to implement this. First I wanted to raise the issue and see if you'd be willing to accept a PR or if you had any feedback.

---

We're currently migrating from Apollo Server to Fastify GQL because of the performance benefits.

Our mobile app makes use of [apollo-link-batch-http](https://www.apollographql.com/docs/link/links/batch-http/) for performance. This batches GraphQL queries which are run within a configurable interval. The request issued is an array of query objects, rather than a singular query object.

In order to support this, the graphql server must be able to merge the query objects and variables prior to the executing the query as per usual.

In order to support this in Fastify GQL:
- [x] Update schema to allow array or object.
- [x] Update handler to merge GraphQL queries if an array is provided.
- [x] Test with batched queries
- [x] Test with normal queries

### PR DESCRIPTION
**Pull request by pscarey**
Fix #185

### CODE REVIEW

**Comment by mcollina** [comment]
Would you mind to add some documentation for batched queries?

**Comment by mcollina** [comment]
code looks great.

**Comment by pscarey** [comment]
Sure will do

**Comment by pscarey** [comment]
Readme has been updated - hope that's OK.

Thanks.

File: lib/routes.js [hunk 1]
```diff
@@ -190,55 +219,21 @@ module.exports = async function (app, opts) {
...
-    let { query } = request.body
-    const {
-      persisted,
-      variables,
-      operationName
-    } = request.body
-
-    if (persisted || onlyPersisted) {
-      query = persistedQueries[query]
+    if (Array.isArray(request.body) && allowBatchedQueries) {
```
Comment (mcollina) [inline lib/routes.js:227]:
```suggestion
    if (allowBatchedQueries && Array.isArray(request.body)) {
```

File: lib/routes.js [hunk 2]
```diff
@@ -7,49 +7,93 @@ const { formatError, GraphQLError } = require('graphql')
...
+  body: {
+    type: allowBatchedQueries ? ['object', 'array'] : 'object',
+    properties: {
+      ...requestProperties,
+      variables: {
+        type: ['object', 'null']
+      }
+    }
+  },
+  // JSON schema isn't allowing ['object', 'array'] on response.
```
Comment (mcollina) [inline lib/routes.js:88]: Have you tried using an `anyOf`? I would prefer to keep the speedup done by the schema.

File: lib/routes.js [hunk 3]
```diff
@@ -190,55 +219,21 @@ module.exports = async function (app, opts) {
...
-      persisted,
-      variables,
-      operationName
-    } = request.body
-
-    if (persisted || onlyPersisted) {
-      query = persistedQueries[query]
+    if (Array.isArray(request.body) && allowBatchedQueries) {
+      // Batched query
+      return Promise.all(request.body.map(async r =>
```
Comment (mcollina) [inline lib/routes.js:229]:
```suggestion
      return Promise.all(request.body.map(r =>
```

This does not need to be async.

File: lib/routes.js (hunk 2 above)
Comment (pscarey) [reply lib/routes.js:88]: I've had a decent look into this after your suggestion, and I can't seem to make it work. Any combination of `anyOf` I tried I got errors compiling the schema.

If you have further suggestions as to where to look for a possible example or e.g. where in the `fast-json-stringify` codebase I might get a lead on what wall I'm hitting, I'm happy to look into it more - am not particularly familiar with JSON Schema though.

**Comment by pscarey** [comment]
I've added the test & made the suggested changes. Thank you. Still stuck as mentioned above with the schema.

File: lib/routes.js (hunk 2 above)
Comment (mcollina) [reply lib/routes.js:88]: Let's land this as is then, we'll investigate in the future.

**Pull request merged**